src/nfa/nfa.cc

```diff
@@ -1354,3 +1562,501 @@ void Mata::Nfa::ShortestWordsMap::update_current_words(LengthWordsPair& act, con
+
+    initialize_pairs_to_process();
+
+    if (lhs.trans_empty() || rhs.trans_empty()) { return; }
+
+    while (!pairs_to_process.empty())
+    {
+        pair_to_process = *pairs_to_process.begin();
+        pairs_to_process.erase(pair_to_process);
+
+        // TODO rewrite this (from previous Vata implementation -- rewrite the algorithm, or the format?)
+
+        compute_for_state_pair();
+    }
+}
+
+void Intersection::compute_for_state_pair()
+{
+    auto this_state_transitions_iter = lhs.transitionrelation[pair_to_process.first].begin();
+    auto other_state_transitions_iter = rhs.transitionrelation[pair_to_process.second].begin();
+    auto this_state_transitions_iter_end = lhs.transitionrelation[pair_to_process.first].end();
+    auto other_state_transitions_iter_end = rhs.transitionrelation[pair_to_process.second].end();
+    // find all transitions that have same symbol for first and the second state in the pair_to_process
+    while (this_state_transitions_iter != this_state_transitions_iter_end
+           && other_state_transitions_iter != other_state_transitions_iter_end)
+    {
+        // first iterator points to transition with smaller symbol, move it until it is either same or further than second iterator
+        if (this_state_transitions_iter->symbol < other_state_transitions_iter->symbol)
+        {
+            while (this_state_transitions_iter != this_state_transitions_iter_end
+                   && this_state_transitions_iter->symbol < other_state_transitions_iter->symbol)
+            {
+                ++this_state_transitions_iter;
+            }
+            if (this_state_transitions_iter == this_state_transitions_iter_end)
+            {
+                break;
+            }
+        }
+        else
+        {
+            // second iterator points to transition with smaller symbol, move it until it is either same or further than first iterator
+            while (other_state_transitions_iter != other_state_transitions_iter_end
+                   && this_state_transitions_iter->symbol > other_state_transitions_iter->symbol)
+            {
+                ++other_state_transitions_iter;
+            }
+            if (other_state_transitions_iter == other_state_transitions_iter_end) { break; }
+        }
+
+        // check both iterators point to the transitions with same symbol
+        if (this_state_transitions_iter->symbol == other_state_transitions_iter->symbol)
+        {
+            compute_for_same_symbols(*this_state_transitions_iter,
+                                     *other_state_transitions_iter);
+
+            ++this_state_transitions_iter;
+            ++other_state_transitions_iter;
+        }
+    }
+}
+
+void Intersection::compute_preserving_epsilon_transitions()
+{
+    initialize_pairs_to_process();
+
+    if (lhs.trans_empty() || rhs.trans_empty()) { return; }
```

Comment:
Is this helpful? I'm thinking that maybe, if this would return, then the while would not be executed anyway, so in other case (i.e. lhs and rhs have trans) would executed two unnecessary tests for emptiness.

Reply:
This is one of the optimizations from the previous Vata2 implementation of intersection (as explained below, the former implementation is not changed at all). Therefore, I have kept such lines there, too.

I will investigate and if I find this unnecessary, I will proceed with removing this check.

Reply:
Previously in the old Vata2, the call to `trans_empty()` was a function with a constant time complexity, calling  `translationrelation.empty()`. However, this is no longer valid. Since the change where each state (even one without any transitions) has to have a place in `transitionsrelation` (`transitionsrelation.at(state)` has to be valid), we can no longer determine the emptiness of transition relation in a constant time. The previous expression now checks whether the automaton has zero states, not zero transitions.

Therefore, these former checks no longer return in constant time for empty transition relations and can be safely removed, exactly as you say.